SNIPS/snip-3668.md

```diff
@@ -0,0 +1,75 @@
+
+This standard is meant to provide an official and audited way to verify the state of Starknet on Ethereum mainnet, and eventually on any EVM-compatible chain. Note the difference is that any chain other than L1 needs to be able to read the Starknet Core Contract's state root in order to check the Merkle inclusion proof's integrity.
+
+## Motivation
+
+There are a few good reasons to read the Starknet state from L1 or another L2.
+
+1. Starknet appchains offering: keystore rollups (few sources: [Keystore Design](https://hackmd.io/@haichen/keystore) and [Scroll's writeup](https://scroll.io/blog/towards-the-wallet-endgame-with-keystore)) and other state dependent appchains (e.g. ENS v2 the appchain) essentially function through multi-chain reading of their state.
+   1. For keystore rollups, L2 or L1 contracts MUST read the user's wallet information in the keystore rollup's state.
+   1. For namespace appchains, L2 or L1 contracts MUST read the user's records in the appchain's state.
+1. Cross-chain capabilities of the Starknet state. Today, other L1 or other L2 onchain applications cannot read the state of Starknet, thus making it impossible to create smooth multi-chain experiences using the Starknet state.
+1. A few Rollup Improvement Proposals (RIPs) are being prepared in Rollcall, the governance body of rollups. Two examples: L1SLOAD and L1STATICCALL enable any rollup contract to respectively (synchronously) read an L1 storage slot and statically call into an L1 contract. These two precompiles would greatly increase the composability of L2s states.
+   1. Note that [Linea](https://github.com/Consensys/linea-ens/tree/main), Scroll, Optimism, Arbitrum and ZKSync Era already implemented this specific ERC for their state: <https://github.com/ensdomains/evmgateway>.
+
+This also opens the floor to further cross-L2 standardizations, as proposed by Vitalik Buterin:
+
+> One of the other idea that we're in the process of figuring out, btw, is moving the de-facto "config" of an L2 to _all_ being callable functions in the L2's core contract on L1.
+
+## Specification
+
+This SRC, if implemented, would result in:
+
+- The deployment of an official Solidity contract able to verify a Merkle inclusion proof for the Starknet state against the Starknet core contract's state root, on Ethereum L1.
+- An off-chain service, so-called "gateway", able to provide inclusion proofs for a specific storage slot.
+- (Optional): The inclusion of a view entrypoint in the Starknet core contract on L1 to verify a Merkle path (This was proposed in Rollcall - the governance body of rollups - by Vitalik Buterin).
+
+The main repositories to check as references are:
+
+- The ENS EVMGateway repository, that lists all the L1 and L2s verifier contracts and their code: <https://github.com/ensdomains/evmgateway>
```

Comment:
> Do the L1 contracts of different chains obey the same interface (say, up to function signatures)?

I'll have to check this in implementations of EIP-3668. I don't know. I hope so otherwise it ruins the purpose of making it an EIP in the first place.